### ISSUE TITLE
Create a new HTML5 converter that maximizes semantic constructs

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#242]
The HTML 5 backend in AsciiDoc is acceptable, but certainly not ideal. It's kind of a stretch to call it HTML 5. We should create an alternative HTML 5 backend that maximizes the use of semantic constructs and keeps the markup to a minimum.

I can cite a few examples of where improvements could be made.
### Paragraphs

I don't see a reason for having paragraphs wrapped in a `<div>`.

```
<div class="paragraph">
  <p>text</p>
</div>
```

This breaks a lot of CSS frameworks, which assume the `<p>` tag will be adjacent to other elements.

It should be handled instead by always using `<p>` tag as the outer element:

```
<p class="paragraph">text</p>
```

and using a nested `<span>` tag for the title of a format paragraph:

```
<p class="paragraph"><strong class="head">Title</strong> text</p>
```

(This is the same approach as used in the [Asciidoctor EPUB3 converter](https://github.com/asciidoctor/asciidoctor-epub3/blob/c313448ecaab2dc0f4f51406226b214058b5bb2d/lib/asciidoctor-epub3/converter.rb#L253-L265)).

### Admonition blocks

Admonition blocks are rendered as a table, which makes styling them very difficult. They should be done instead as layered `<div>` tags that can accommodate a wider range of layouts.
### Sidebars

`<aside>` element. Nuff said.
### Header and footer

`<header>` and `<footer>` elements. Nuff said.
### Sections

`<section>` element. Nuff said.
### Block images

`<figure> and`<figcaption>` element. Nuff said.
### Content

`<article>` element if doctype is article, or perhaps the `<main>` element.
### Dates and times

`<time>` element. Nuff said.

**Comment by Shadowfiend** [comment]
What's the status here, out of curiosity? Is there something I can do to help? The current asciidoctor output is rather painful (if fully functional :)), and I'd love to get a proper HTML5 conversion off the ground quickly. It's not really clear from this issue (and the ones linked from it) if there's been progress made already, if there's already a place (repo or subdirectory) where that progress might take place, etc.

**Comment by mojavelinux** [comment]
One of my goals for 2017 is to see that this happens.

A lot of the thinking for this issue ended up happening as part of developing both the [EPUB3 converter](https://github.com/asciidoctor/asciidoctor-epub3/blob/master/lib/asciidoctor-epub3/converter.rb) and [Bespoke converter](https://github.com/asciidoctor/asciidoctor-bespoke/tree/master/templates/slim). Both of those converters use semantic HTML. While the HTML for the HTML5 converter certainly has not been decided or settled on, those converters serve as prior art and get us much further down the path.

There are three parts to this issue:

1. Decide on the HTML structure, in general and for each node
2. Write the converter to implement that structure
3. Figure out when and how this becomes the default converter (that may entail creating a new default stylesheet)

Where people can help on most is (1). Take a node and spec out the HTML. In other words, what HTML should it make? We need decisions.

For example, how should a paragraph node be converted? It's necessary to account for all existing settings/attributes. See https://github.com/asciidoctor/asciidoctor/blob/7282b0a5864df2f587b5a5a8f3cda10f71ed5547/lib/asciidoctor/converter/html5.rb#L701-L715. How about a sidebar?

**Comment by mojavelinux** [comment]
To answer your question more directly about what has been done...I'd say that what has been done is that proposals have been made in the EPUB3 and Bespoke converters. We need to pick out what we want from those and add any additional semantics that are needed for desktop HTML. You can submit proposals as custom Slim templates or as a Ruby-based converter in a fork of the project.

One of the big discussions is going to be on the outer framing, which is what the document node writes. That's where the biggest limitations to styling are currently introduced. It's a very strange, flat structure where the header, content, and footer are all sibling divs with no common parent.

**Comment by rockyallen** [comment]
This may be off-topic, but could the new converter produce well-formed XML? The current output isn't, which makes it less convenient to post-process with XML tools.

**Comment by mojavelinux** [comment]
@rockyallen That's a setting. You just need to set the backend to xhtml and you get compliant XML / XHTML.

There are as many arguments for HTML (not XML) as there are for XHTML (XML). That's why we offer it as an option. In a sense, the HTML converter is really two converters in one.

Here's where it's covered in the docs. Feel free to augment. http://asciidoctor.org/docs/user-manual/#xhtml

**Comment by mojavelinux** [comment]
The semantic converter will do the same. We have be able to produce both variants as they are simply different things.

**Comment by fap-** [comment]
Since this resource seems useful and it hasn't been posted so far:
http://a11yproject.com/posts/aria-landmark-roles/
http://a11yproject.com/checklist.html

**Comment by jirutka** [comment]
Just three weeks ago I’ve returned back to my Semantic HTML5 converter just a week ago [asciidoctor-html5](https://github.com/jirutka/asciidoctor-html5) and it’s almost finished. I focus on accessibility too, but ARIA part is still in progress. I’m currently researching it.

I found WAI-ARIA very helpful https://www.w3.org/TR/dpub-aria-1.0/.

**Comment by mojavelinux** [comment]
I've discovered a limitation with the table caption. Assigning it a width does not work reliably across browsers. Firefox computes the width relative to the table parent. Chrome and Edge computes the width relative to table itself. To mitigate this problem, I think the table should be wrapped in a figure element and the figcaption used for the block title. I think this is closer to the semantics we are shooting for anyway.

**Comment by mojavelinux** [comment]
* Use figure and figcaption for HTML5 image output

**Comment by mojavelinux** [comment]
* Use figure and figcaption for quote block (or somehow get the attribution to not be floating outside of the figure when rendered without a stylesheet).

**Comment by jirutka** [comment]
> Use figure and figcaption for quote block (or somehow get the attribution to not be floating outside of the figure when rendered without a stylesheet).

I don't think that it'd be correct use of figure/figcaption (as I remember from research I did few years ago) and it's not even necessary. [Here](https://github.com/jirutka/asciidoctor-html5s/blob/master/test/examples/html5/quote.html) you can see how I designed it in html5s converter (which goal is to be semantically correct).

**Comment by znerd** [comment]
Is this backend already available in the standard distribution?

Would it make sense to make this backend an _experimental_ feature of Asciidoctor? E.g.

```
asciidoctor -b html5s-experimental
```

Then it's as easy as possible to start trying this out, but on the other hand it's very explicit that this is experimental.

Secondly, would it make sense to add a test case for a significant AsciiDoc file and an associate expected HTML output file? Then this can be shaved and shaved as needed, potentially with PRs from contributors.

**Comment by jirutka** [comment]
@znerd It’s not, but you can use [asciidoctor-html5](https://github.com/jirutka/asciidoctor-html5) (and suggest improvements). It’s trivial to install (`gem install asciidoctor-html5s -v 0.1.0.beta.11`, no dependencies except for asciidoctor) and use (just `asciidoctor -r asciidoctor-html5s -b html5s`), supports both Asciidoctor and Asciidoctor.js, easy to modify (uses Slim templates with clean separation of complex logic into ruby helpers) and fully tested using [Asciidoctor::DocTest](https://github.com/asciidoctor/asciidoctor-doctest).

**Comment by znerd** [comment]
@jirutka Thanks for the recommendation and info!

@mojavelinux What is the strategy (if any is decided yet) regarding the semantic HTML backend?
- a new backend could be started from scratch,
- or the existing `asciidoctor-html5s` backend (by @jirutka ) could be integrated as-is,
- or something in the middle?

**Comment by mojavelinux** [comment]
Work will begin integrating the semantic HTML converter into core shortly after the upcoming 2.0.0 release. Of course, we'll build on the amazing work that has already been done. But there will still be tough decisions to make.

The very rough goal will be to introduce a new converter in the 2.x series and make it the default in 3.0.0.

In the meantime, we need an interim name (before it becomes html). One idea is to name it html51 (or html52) since those names aren't taken. I don't love it though since we're not really developing for a specific HTML version. Perhaps "s-html5" or "sem-html5". Another approach would be "htmlnext". (I don't really like the "s" after the name).

One thing I can say for certain is that, like the current HTML converter, it will not use any external libraries. However, I'll very likely make greater use of functions to avoid code duplication and make extending it easier.

**Comment by mojavelinux** [comment]
> Would it make sense to make this backend an experimental feature of Asciidoctor? 

Regardless of whether it's experimental or not, it still needs to exist ;) I haven't had the resources to start working on it up until now, but that's about to change.

**Comment by mojavelinux** [comment]
Of course, there will be tests. Unlike other features in Asciidoctor, this feature has to be done very carefully. Part of the problem with the existing HTML converter is that it was thrown together haphazardly. This new converter needs to be done right from the start...and each element has to be carefully selected to give us both the semantics we want and be easy to style. (Fortunately, we have a lot of ideas to draw from). Having worked extensively with the existing HTML, I'm keenly aware of what that HTML needs to be. But that doesn't mean I don't still have a lot of questions. It's just that those questions have to be taken as part of a whole.

**Comment by mojavelinux** [comment]
When a role is assigned to a section title, we should think whether that gets added to the enclosing section or directly on the heading. Or, perhaps, the role should be added to the heading and also added to the section, but with a fixed prefix or suffix (e.g., `rolename-sect`). This came from #1239.

**Comment by mojavelinux** [comment]
Block and inline attributes that begin with `data-` (i.e., data attributes) should be passed directly to the container element in the HTML output (1-to-1 mapping). This provides a simple way to pass data from the AsciiDoc document directly to the HTML output.

**Comment by mojavelinux** [comment]
Since it's never a good time, now's as good a time as ever. I would seriously like to start entertaining the idea of starting work on the new HTML converter so we can begin clearing away some of these issues.

Let's not get hyper-focused on what to call it yet. For now, we can just give it the working name of html5.2 or semhtml5 and we can rename it before it gets finalized if needed.

The focus should be a) semantic HTML, such as `<section>` tags for the document sections, etc, consistent structure for block and styled paragraphs, b) code reuse (which is a major issue with the current converter), and c) modern integrations like Font Awesome 5. When it's all said and done, we want people to like the HTML that Asciidoctor produces. And we should be proud of it too.

We have plenty of raw material to build on. This gives us plenty of room to experiment without affecting any existing uses of Asciidoctor.

**Comment by luvarqpp** [comment]
A bit off-topic, please have on mind (when architecture of code forms) also integration of external resources and ability to opt-out from them. Ability to "download them and embed/save in resulting files" would be welcome.
I would like to be able to generate **self-contained html** documentation conforming with very strict [CSP](https://content-security-policy.com/) to serve whole documentation from "self" and inline if convenient. I mean mainly nicely documented (see https://github.com/asciidoctor/asciidoctor.org/issues/578 and other) option to generate such documentation. (I have found this "weird" think in hard way. My [default jenkins set-up](https://wiki.jenkins.io/display/JENKINS/Configuring+Content+Security+Policy) had served my documentation in different style than I was used when loaded from local build (no "NOTE" icons, no sidebar, e.t.c.)

**Comment by mojavelinux** [comment]
@luvarqpp where it pertains to this converter, I agree. However, the general topic needs to be tracked in a dedicated issue. It's a bit off topic here and could thus serve to slow down the progress on it. Let's continue the discussion in #578 and mention specifics here if they pertain to the converter itself.

**Comment by Mogztter** [comment]
> Of course, there will be tests. Unlike other features in Asciidoctor, this feature has to be done very carefully. Part of the problem with the existing HTML converter is that it was thrown together haphazardly. This new converter needs to be done right from the start...and each element has to be carefully selected to give us both the semantics we want and be easy to style. (Fortunately, we have a lot of ideas to draw from). 

I think it could be worth to open a pull request for each convert method so we can carefully review it, explain/document why we chose this semantic construct and add the corresponding tests (I believe that each convert method can be tested independently, right?).

> Having worked extensively with the existing HTML, I'm keenly aware of what that HTML needs to be. But that doesn't mean I don't still have a lot of questions. It's just that those questions have to be taken as part of a whole.

> The focus should be a) semantic HTML, such as `<section>` tags for the document sections, etc, consistent structure for block and styled paragraphs, b) code reuse (which is a major issue with the current converter), and c) modern integrations like Font Awesome 5. When it's all said and done, we want people to like the HTML that Asciidoctor produces. And we should be proud of it too.

Maybe we should write it down in an architectural decision record (ADR).
Implementation wise the new semantic converter will be written in pure Ruby (ie. it will not use a template library) ?

> b) code reuse (which is a major issue with the current converter)

👍 
A good candidate might be the Table Of Content, I found 4 occurrences in the current HTML5. Do you think it makes sense to introduce methods to create HTML elements, like `create_section` ?

**Comment by john-cj** [comment]
@mojavelinux

> I've discovered a limitation with the table caption. Assigning it a width does not work reliably across browsers. Firefox computes the width relative to the table parent. Chrome and Edge computes the width relative to table itself.

It was April 2018, before Firefox 57, and since Firefox has moved (is moving?) from Gecko to Servo, this issue is possibly doesn't exist any more.

> Decide on the HTML structure, in general and for each node

Few ideas from Apple documentation of the Swift language:

https://web.archive.org/web/20180420193345/https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/TheBasics.html#//apple_ref/doc/uid/TP40014097-CH5-ID309

1.  HTML structure.

    ```html
    <!--
    The structure:

    * abc
      * aaa
        * foo
      * bbb
        * bar
    -->

    <article class="chapter">
        <h2 class="chapter-name">abc</h2>
        <section>
            <!--
            The 1st section in the chapter.
            It doesn't has its own heading;
            instead, it has chapter heading
            right about itself.
            -->
            <p>...</p>
            <p>...</p>
            <p>...</p>
        </section>
        <section>
            <h3 class="section-name">aaa</h3>
            <p>...</p>
            <p>...</p>
            <p>...</p>
            <section>
                <h3 class="section-name">foo</h3>
                <p>...</p>
                <p>...</p>
                <p>...</p>
            </section>
        </section>
        <section>
            <h3 class="section-name">bbb</h3>
            <p>...</p>
            <p>...</p>
            <p>...</p>
            <section>
                <h3 class="section-name">bar</h3>
                <p>...</p>
                <p>...</p>
                <p>...</p>
            </section>
        </section>
    </article>
    ```

2. Code blocks with line numbering and the option to use each line number as a reference. ("See Listing 5, line 5.")

**Comment by mojavelinux** [comment]
@john-cj These are useful references. Thanks.

**Comment by cup** [comment]
https://github.com/asciidoctor/asciidoctor/issues/1087

**Comment by gdamore** [comment]
Just as a note: I've been working on the libasciidoc project, and we have endeavored to generate mostly the same output as asciidoctor, even when that output has some ... less than ideal idiosyncrasies.

I'd really like to see this stuff cleaned up, and I had many of these same thoughts as here, but we simply didn't want to break 1:1 compatibility.  (One significant reason for this is the consideration of CSS stylesheets that might depend on these idiosyncrasies.)

**Comment by Teknomancer** [comment]
I'm not entirely sure if this particularly was mentioned before, so just mentioning it in case it wasn't.

Asciidoctor (2.0.10) seems to generate inline styles (CSS) when using tables and outputting as HTML namely for column width (and possibly other things?

e.g (adoc):
```
[options="header",stripes=hover]
|===
| File | Version | OS  | Size
| eg1 | 1.0 | Windows | 1 MB
|===
```

CSS:
```
<table class="tableblock frame-all grid-all stripes-hover stretch">
<colgroup>
<col style="width: 25%;"/>
<col style="width: 25%;"/>
```
The inline `style` in the `col` tag leads to Content-Security-Policy violations.

Just adding it here in case we are building some list of things that need to be cleaned of inline CSS.

**Comment by mojavelinux** [comment]
Thanks for pointing this out. In this case, the styles are necessary, unless we switch to using the width attribute instead. Aside from those two options, there's no other way I know of to specify an arbitrary width on a column, which is required to support AsciiDoc tables properly.

As a side note, I am staunchly opposed to avoiding the use of the style attribute on principle. I don't know who decided this was a security violation, but I can only conclude that person/group has a limited understanding of HTML. It's an absurd restriction and I reject it.

**Comment by Teknomancer** [comment]
I totally agree with you that this is somehow a security issue seems quite over the top. I'm not a security expert, but considering inline CSS as a security issue seems rather extreme. Anyway...

Would adding an Asciidoctor table option be possible to prevent the `style=width` from being generated? Because I didn't specify any width in the .adoc and I wouldn't mind automatic width in simple tables, i.e. the behavior as it would when no `width` is specified in `col`.

**Comment by mojavelinux** [comment]
We cannot modify the logic for what determines when a column gets a width or not. That's well established in AsciiDoc (and something that can be controlled).

I do wonder whether the right solution is to use the width attribute. I cannot remember at the moment why we stopped using it. I think I was trying to be consistent with other places where the style attribute is required. The change would need to be thoroughly tested that it is supported in the same way in modern browsers. But if it is, it seems like a reasonable change to make.

**Comment by Teknomancer** [comment]
For those wondering if there's a way to mitigate this without resorting to `unsafe-inline` in your CSP, there is `style-src-attr` (https://groups.google.com/a/chromium.org/g/blink-dev/c/uauMQQQ7fAI?pli=1)

Using `style-src-attr 'unsafe-inline'` in your CSP header instead of `style-src 'unsafe-inline'`, should be a much safer solution. This might be a good interim solution.

However, I'm not certain how good the browser support for `style-src-attr` is. According to https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src-attr (it is fully supported on Chrome, Edge but not on Firefox). It's also a better solution than rejecting inline styles altogether (in case where you are not in control of the CSP headers your webserver emits and just hosting content using Asciidoctor) as it would work at least on a few browsers.

**Comment by mojavelinux** [comment]
While working on issues related to table cell borders, I had an idea for the semantic HTML5 output. If a cell spans columns or rows, and that cell ends up touching the right or bottom boundary of the table, the converter should add a class to indicate which boundary it touches. This is necessary since cells that span columns or rows aren't represented in the column or row they expand into in the HTML. That means they don't get matched by the CSS selector used to control the cell borders. So we need a hint. And that hint can come in the form of a class that the HTML converter adds. See #3447.

**Comment by Mogztter** [comment]
> I do wonder whether the right solution is to use the width attribute. I cannot remember at the moment why we stopped using it.

The `width` attribute is [deprecated](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/col#attr-width) but still widely supported.
This attribute is no longer part of the list of attributes that may be specified on `<col>` elements in the [HTML 5 specification](https://www.w3.org/TR/html52/tabular-data.html#the-col-element) (same thing in the [HTML Living Standard](https://html.spec.whatwg.org/multipage/tables.html#the-col-element)). It was part of the [HTML 4 specification](https://www.w3.org/TR/html401/struct/tables.html#h-11.2.4.2).

Having said that the alternatives are not great...

**Comment by mojavelinux** [comment]
@Mogztter Thanks for pointing that out. I knew there had to be a decent reason ;)

**Comment by Teknomancer** [comment]
Would using lists `ol` rather than `div` tags for listing footnotes fall within the scope of this defect?

I understand there might be advanced cases where footnotes might not be numbered (custom numbering or text?) but for the default case where Asciidoctor auto-generates numbers, the semantic backend might be able to use `ol` for this purpose.

e.g. footnotes are currently generated like this:
```
<div id="footnotes">
  <hr/>
  <div class="footnote" id="_footnotedef_1">
  <a href="#_footnoteref_1">1</a>. footnote text.</div>
</div>
```
The current solution makes styling footnotes a tad tricky. For instance, if you want to introduce padding between the footnote 
number (plus period) and the text that follows it.

Now if `ol` were to be used, we might not be able to use the number as the backlink since that's part of the `li` marker, but perhaps a ↩ or `^` needs to be added to the footnote text (before or after) as the backlink instead?

Hmm... perhaps this might be too drastic a change that breaks compatibility.

**Comment by mojavelinux** [comment]
I've created the feature/html-converter-next branch to start accepting the code for this change. Once the converter is fully functional, we'll fold that into the main branch to be included in a 2.x release as a technology preview (backend name TBD). The plan is to make this the default HTML converter in a future major release TBD. But we'll need lots of testing before then.

**Comment by znerd** [comment]
That branch is 344 commits behind master. I guess it makes sense to at least keep it more-or-less in sync…

**Comment by john-cj** [comment]
There is a box above that says "38 hidden items Load more…"

I try to click it, but nothing happens, and the browser console shows error 500 because of tying to access

https://github.com/asciidoctor/asciidoctor/issues/242/partials/load_more?after_cursor=Y3Vyc29yOnYyOpPPAAABYae9WmACqTU5NDYwNjY4Mg%3D%3D&before_cursor=Y3Vyc29yOnYyOpPPAAABb04RGQgAqTU2OTQ0NjMwOQ%3D%3D

Is it just me or other people have the same issue?

**Comment by mojavelinux** [comment]
@john-cj Please report that to GitHub. We have no control over this interface.

**Comment by 89z** [comment]
https://github.com/asciidoctor/asciidoctor/issues/1087

**Comment by codemaestro** [comment]
Hi all, where s this initiative at after 10 years?

**Comment by redbow-kimee** [comment]
For anyone looking to help out with the semantic HTML converter, choose a missing function between these commands, check that nobody else has an open merge for it, then implement the missing function, and add some tests.

## Find a missing convert function
These functions convert the asciidoc syntax tree into like html strings, implementing each feature. `lib/asciidoctor/converter/html5.rb` implements the "current" converter which uses html tables for everything. The new converter, which uses semantic html, is (unsurprisingly) in   `lib/asciidoctor/converter/semantic_html5.rb`

These commands should help you compare them.

```shell
git checkout feature/html-converter-next
grep 'def convert_' lib/asciidoctor/converter/html5.rb | sort | awk '{ print $2 }' FS=' '
grep 'def convert_' lib/asciidoctor/converter/semantic_html5.rb | sort | awk '{ print $2 }' FS=' '
```
## Prevent duplication of work
Unfortunately, there isn't a good listing of what's in progress, although I do know of https://github.com/asciidoctor/asciidoctor/pull/4321 and https://github.com/asciidoctor/asciidoctor/pull/4322. Scan through the open merge requests, and if none of the titles mentioning "semantic html5" mention the feature you're planning to fix, then you should be good to go.

Actually, since to my knowledge all the open MRs to do with this have "semantic HTML 5" (exact spacing and capitalization) in their titles, this link should help search for them https://github.com/asciidoctor/asciidoctor/pulls?q=is%3Apr+is%3Aopen++%22semantic+HTML+5%22+in%3Atitle+

## Implementing and testing.
Write some code, I can't necessarily help you there. But when you're ready to test here's how. The semantic HTML converter uses adoc inputs with expected outputs for testing. These can all be found in `test/fixtures/semantic-html5-scenarios/`. Add a new adoc file there (or several, to impart more constraints on your code).

Then run these commands to run the tests.

```shell
# Run your test, printed to console until you're happy with it.
./bin/asciidoctor --backend semantic-html5 --embed -o test/fixtures/semantic-html5-scenarions/${test_name}.adoc
# Save the output of your test.
./bin/asciidoctor --backend semantic-html5 --embed -o test/fixtures/semantic-html5-scenarios/${test_name}.html test/fixtures/semantic-html5-scenarios/${test_name}.adoc
# run all the tests tests for the semantic html converter.
ruby test/semantic_html5_converter_test.rb
```

## Make a merge request
Now make a merge requests and make sure to put "semantic HTML 5" (exact spacing and capitalization) in your issue title, so the above search link continues to work.

**Comment by akhayyat** [comment]
Is the new html converter still happening, or is it abandoned?

**Comment by mojavelinux** [comment]
Yes, it's still happening. It's just not a priority at the moment over other matters. If you would like to discuss it, or volunteer to help, please join us in the chat at https://chat.asciidoctor.org.

**Comment by ggrossetie** [comment]
This discussion has been going for a long time, and I think part of the reason it's hard to converge is that most of the semantic decisions (paragraph wrapping, admonitions, sidebars, footnotes, etc.) are being debated in the abstract. There's no working implementation people can actually try in a real project. And given how widely Asciidoctor is used, there's understandable pressure to get every decision right before anything lands in the default backend, which slows things down further.

I'd like to propose using **Asciidoctor.js as a fast-iteration testbed** for this work, explicitly labeled as experimental. I've discussed this with Dan beforehand to make sure it supports the existing effort rather than competing with it.

Concretely, I'm going to implement a `semantic-html5` backend in Asciidoctor.js:

- It will start as a **direct port of the `feature/html-converter-next` branch**, then be completed to cover all remaining node types (lists, tables, admonitions as `<aside>`, quotes/verses with `<footer>` attribution, TOC as `<nav>`, footnotes as `<section role="doc-endnotes">`, video/audio, and a standalone layout using `<header>`/`<main>`/`<footer>`).
- The **scenario fixtures from that branch will be the shared source of truth**: the existing `.adoc`/`.html` pairs will be kept and extended as coverage grows, so every decision validated on the JS side stays directly comparable with the Ruby branch.
- Every element mapping and every deliberate deviation from the current `html5` output will be documented in a design doc, so nothing gets decided silently.
- It will ship with a dedicated default stylesheet (light/dark, CSS custom properties) so people can evaluate real rendering out of the box, not just raw markup.

The point of shipping it as an experimental backend is the feedback loop: anyone will be able to try it immediately in real tooling (VS Code extension, documentation sites, CLI) and give concrete feedback based on actual rendering rather than opinions on markup snippets. The backend will be clearly documented as a prototype. Its output may change in any release, and it is *not* a candidate to replace the default `html5` backend. Once a set of decisions proves solid through real-world usage, the intent is to port them back to the Ruby implementation here.

I've opened a tracking issue on the Asciidoctor.js side: https://github.com/asciidoctor/asciidoctor.js/issues/1852. Feedback on specific markup decisions is very welcome there (or here, for the decisions that belong to this discussion).
